src/util/sss_regexp.c

```diff
@@ -0,0 +1,215 @@
+/*
+    SSSD
+
+    Authors:
+        Tomas Halman <thalman@redhat.com>
+
+    Copyright (C) 2018 Red Hat
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "util/sss_regexp.h"
+#include <string.h>
+
+#define SSS_REGEXP_OVEC_SIZE 30
+#define SSS_REGEXP_ERR_MSG_SIZE 120 /* 120 is recomended by pcre2 doc */
+
+#ifdef HAVE_LIBPCRE2
+/*
+ * sss_regexp with pcre2
+ */
+struct _sss_regexp_t {
+    pcre2_code *re;
+    unsigned char *error_msg;
+    pcre2_match_data *match_data;
+    char *matched_string;
+};
+
+int sss_regexp_pcre2_destroy(sss_regexp_t *self)
+{
+    if (self->re) pcre2_code_free(self->re);
+    if (self->error_msg) free (self->error_msg);
+    if (self->match_data) pcre2_match_data_free(self->match_data);
+    if (self->matched_string) pcre2_substring_free((PCRE2_UCHAR *)self->matched_string);
+    return 0;
+}
+
+void sss_regexp_pcre2_compile(sss_regexp_t *self,
```

Comment:
I wanted the change to be as small as possible -- this is how pcre/pcre2 does it, so I keep it here. There are actually three output error parameters errorcode, error message and error position.

Reply:
OK, but the caller of `sss_regexp_new` only checks the return value of `sss_regexp_new` which is the result of allocation of `sss_regexp_t`. So if you can allocate the structure, but then the regex parsing fails, noone currently checks the error. Wouldn't it be better to also e.g. check the validity of `self->re` inside `sss_regexp_t` and set some fallback error code and value even if `sss_regexp_pcre2_compile` doesn't set one? Also, the failure to compile the regex must be somehow passed up to the caller, currently the caller only checks the validity of the allocation.

Reply:
Ahh, I finally see what you mean. Caller fixed on several places to check the errorcode instead of return value. Errorcode is always set, even on allocation failure.